BZ-1901643: Updated Monitoring VM health assembly

sjhala-ccs opened the pull request:
This PR addresses https://bugzilla.redhat.com/show_bug.cgi?id=1901643

Updated the Monitoring VM health assembly per doc enhancement suggestions in the BZ.

Applies to 4.6 and later.

Preview build: https://deploy-preview-34709--osdocs.netlify.app/openshift-enterprise/latest/virt/logging_events_monitoring/virt-monitoring-vm-health?utm_source=github&utm_campaign=bot_dp

Requesting SME review from @stu-gott @kwiesmueller 
cc: @aburdenthehand

aburdenthehand commented on modules/virt-about-liveness-readiness-probes.adoc:
```diff
@@ -6,7 +6,14 @@
 
-_Readiness probes_ tell services and endpoints that the virtual machine instance
-is ready to receive traffic from services. If readiness probes fail, the virtual machine instance is removed from applicable endpoints until the probe recovers.
+A _readiness probe_ determines whether a VMI is ready to accept service requests. If the probe fails, the VMI is removed from the list of available endpoints until the probe recovers.
```
Nit of all nits (brace yourself for pettiness): 
Since we say 'liveness and readiness probes' in title and para and examples, should liveness come before readiness here?

sjhala-ccs replied:
Done

aburdenthehand reviewed (commented):
A couple of nitty nits and some questions stemming more from curiosity. 
Much improved - thank you

kwiesmueller commented on modules/virt-define-http-liveness-probe.adoc:
```diff
@@ -4,74 +4,36 @@
+    initialDelaySeconds: 120 <1>
+    periodSeconds: 20 <2>
+    httpGet: <3>
+      port: 1500 <4>
```
I think we should add more details on the probe, an example from the official K8s docs:
```yaml
  livenessProbe:
      httpGet:
        path: /healthz
        port: 8080
        httpHeaders:
        - name: Custom-Header
          value: Awesome
      initialDelaySeconds: 3
      periodSeconds: 3
```
I'd at least add path as well.

sjhala-ccs replied:
Added

kwiesmueller commented on modules/virt-template-vmi-probe-config.adoc:
```diff
@@ -0,0 +1,58 @@
+      initialDelaySeconds: 120
+      periodSeconds: 20
+      tcpSocket: 
+        host: <host_IP>
```
I think in this case the host is not needed. It should always default to the Pod IP. So only port should be specified.
In this case probably `port: 1500` as well to make this work as it will then try to TCP the same endpoint that provides the readinessProbe.
In general, I'd honestly not include the livenessProbe. Still think it's unlikely for people to use it a lot with VMIs and if the user tries on a slow system their VM might just get stuck in a crash loop.

sjhala-ccs replied:
Okay, I have removed that portion from the template.

sjhala-ccs commented on modules/virt-define-http-liveness-probe.adoc:
```diff
@@ -4,74 +4,36 @@
 [source,terminal]
 ----
-$ oc create -f <file name>.yaml
+$ oc create -f <file_name>.yaml
```
@kwiesmueller Regarding Andrew's first comment, I remember we talked about using the VM config instead of the VMI? Would that make more sense? 

@aburdenthehand As far as I understand, you cannot edit an existing VMI. The probe information would be added when first creating the VMI.

sjhala-ccs commented on modules/virt-about-liveness-readiness-probes.adoc:
```diff
@@ -6,7 +6,14 @@
+
+* HTTP GET: With this test, the probe determines the health of the VMI by using a web hook. The test is successful if the HTTP response code is between 200 and 399. You can use an HTTP GET test with applications that return HTTP status codes when they are completely initialized.
+
+* TCP socket: With this test, the probe attempts to open a socket to the VMI. The VMI is only considered healthy if the probe can establish a connection. You can use a TCP socket test with applications that do not start listening until initialization is complete.
```
TCP socket test is for both types of probes. Yeah, the example shows only HTTP for readiness. TCP socket example for readiness probe would be the same as TCP liveness probe. Maybe I can point that out in the doc.

sjhala-ccs commented:
@kwiesmueller I have incorporated your comments and updated the content. Since we are downplaying liveness probes, I have moved the readiness probe example to the top and swapped the TCP test example for the liveness probe with one for the readiness probe, assuming they would be similar.

Preview link: https://deploy-preview-34709--osdocs.netlify.app/openshift-enterprise/latest/virt/logging_events_monitoring/virt-monitoring-vm-health?utm_source=github&utm_campaign=bot_dp

Can you please take a look, and if the changes look good, approve them? I would like to move this bug to QA soon. Thank you.

kwiesmueller commented on modules/virt-define-http-readiness-probe.adoc:
```diff
@@ -0,0 +1,42 @@
+spec:
+  readinessProbe:
+    httpGet: <1>
+      port: 1500 <2>
```
I think we should keep this similar to the one below and also show/explain the full example (including path).
So the examples are consistent.

sjhala-ccs replied:
Added the path and httpHeaders fields similar to the liveness probe example

kwiesmueller commented on modules/virt-define-http-readiness-probe.adoc:
```diff
@@ -0,0 +1,42 @@
+    successThreshold: 3 <7>
+----
+<1> The HTTP GET request to perform to connect to the VMI.
+<2> The port of the VMI that the probe queries. In the above example, the probe queries port 1500. The VMI installs and runs a minimal HTTP server on port 1500 via cloud-init.
```
`The VMI installs and runs a minimal HTTP server on port 1500 via cloud-init.`
This is not true in the example as the cloud-init part is missing. It's only shown below.

sjhala-ccs replied:
Will remove that line

kwiesmueller commented on modules/virt-template-vmi-probe-config.adoc:
```diff
@@ -0,0 +1,52 @@
+
+[source,yaml]
+----
+  apiVersion: kubevirt.io/v1alpha3
```
Were you able to verify this example works or do you need help with that?

sjhala-ccs replied:
The template is based on the existing documentation example: https://docs.openshift.com/container-platform/4.8/virt/logging_events_monitoring/virt-monitoring-vm-health.html#virt-define-readiness-probe_virt-monitoring-vm-health

But it would good to verify that it works. Can you please help with that? Thanks.

kwiesmueller reviewed (changes requested):
Thank for the update!

kwiesmueller commented on modules/virt-template-vmi-probe-config.adoc:
```diff
@@ -0,0 +1,52 @@
+    terminationGracePeriodSeconds: 0
+    volumes:
+    - name: containerdisk
+      registryDisk:
```
```suggestion
      containerDisk:
```

sjhala-ccs replied:
Fixed

kwiesmueller commented on modules/virt-template-vmi-probe-config.adoc:
```diff
@@ -0,0 +1,52 @@
+          bootcmd:
+            - setenforce 0
+            - dnf install -y nmap-ncat
+            - systemd-run --unit=httpserver nc -klp 1500 -e '/usr/bin/echo -e HTTP/1.1 200 OK\\n\\nHello World!'
```
When running the VM the probe fails with
```
Readiness probe failed: Get "http://10.129.2.34:1500/": net/http: HTTP/1.x transport connection broken: unexpected EOF
```
Going into the VM and using curl the connection looks fine.
Maybe @phoracek knows if the HTTP probe is right this way?

phoracek replied:
Could it be that firewall is blocking that port? Does calling curl from the outside get to the VM?

kwiesmueller replied:
I tried again today and it was fine. Might have been that or just my impatience.

sjhala-ccs commented:
@kwiesmueller Can you please take a look and approve the changes if they look good? Thank you.

kwiesmueller commented:
/lgtm
Thanks, I had another look.
Maybe @stu-gott wants to have another look from his team.

sjhala-ccs commented:
Requested QE review from @gouyang

gouyang commented:
LGTM

ousleyp commented on modules/virt-define-http-liveness-probe.adoc:
```diff
@@ -4,74 +4,41 @@
-= Define an HTTP liveness probe
+= Defining an HTTP liveness probe
+
+Define an HTTP liveness probe by setting the `spec.livenessProbe.httpGet` field of the virtual machine instance (VMI) configuration. Both HTTP and TCP tests for liveness probes can be defined in the same way as readiness probes. This procedure configures a sample liveness probe with an HTTP GET test.
```
I might rephrase this sentence to the active voice: "Both HTTP and TCP tests for liveness probes can be defined in the same way as readiness probes."

ousleyp commented on modules/virt-define-http-liveness-probe.adoc:
```diff
@@ -4,74 +4,41 @@
 ----
-+
-. Create the virtual machine instance by running the following command:
+<1> The time, in seconds, after the VMI starts before the liveness probe is initiated.
```
Suggestion: I might rephrase the `<1>` callout so that it resembles the `<2>` callout (referring to a delay between events).

sjhala-ccs replied:
Keeping this as is for consistency with the other modules

ousleyp commented on modules/virt-define-http-liveness-probe.adoc:
```diff
@@ -4,74 +4,41 @@
 [source,terminal]
 ----
-$ oc create -f <file name>.yaml
+$ oc create -f <vmi-config>.yaml
```
The other two modules use `<file_name>.yaml`; I think it would be best if they were all the same. Outside of that, this should use an underscore instead of a hyphen.

sjhala-ccs replied:
Fixed

ousleyp commented on modules/virt-define-tcp-readiness-probe.adoc:
```diff
@@ -0,0 +1,39 @@
+
+.Sample readiness probe with a TCP socket test
+[source,yaml]
+----
```
(applies to all snippets) according to the OCP docs contribution guide, include ellipses to indicate that the file continues: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

Optional, but encouraged: include the beginning of the YAML so that the `kind:` field is visible.

sjhala-ccs replied:
Added ellipses at the beginning and end of the code block for all snippets

ousleyp commented on modules/virt-template-vmi-probe-config.adoc:
```diff
@@ -0,0 +1,52 @@
+
+[source,yaml]
+----
+  apiVersion: kubevirt.io/v1alpha3
```
This whole example file is indented by two spaces. I'm not sure if that actually breaks anything, but it is probably best to remove those extra spaces.

sjhala-ccs replied:
Removed the extra space

ousleyp commented on virt/logging_events_monitoring/virt-monitoring-vm-health.adoc:
```diff
@@ -5,10 +5,15 @@ include::modules/virt-document-attributes.adoc[]
 
-Use this procedure to create liveness and readiness probes
-to monitor virtual machine health.
+A virtual machine instance (VMI) can become unhealthy due to transient issues such as temporary connectivity loss, deadlocks, or problems with external dependencies. A health check periodically performs diagnostics on a VMI by using a combination of the readiness and liveness probes.
```
I don't think you need "temporary" here, since "transient" applies to the whole list.

I am confused by this sentence: "A health check periodically performs diagnostics on a VMI by using a combination of the readiness and liveness probes." 
- Is a 'health check' in this context something other than what you configure in this assembly? 
- "...by using a combination of the readiness and liveness probes" makes it sound like you have to use both, but the assembly makes it seem like you can use one or the other.

sjhala-ccs replied:
Removed "temporary" 

A "health check" means the same thing as configuring the probes as described in the assembly. I changed the second sentence to "by using any combination of the readiness and liveness probes".

ousleyp reviewed (commented):
This mostly looks great, I just have a few minor nits and suggestions. Thanks, Shikha!

sjhala-ccs commented on modules/virt-about-readiness-liveness-probes.adoc:
```diff
@@ -0,0 +1,19 @@
+
+A _liveness probe_ determines whether a VMI is responsive. If the probe fails, the VMI is deleted and a new instance is created to restore responsiveness.
+
+You can configure readiness and liveness probes by setting the `spec.readinessProbe` and the `spec.livenessProbe` fields of the `VirtualMachineInstance` object respectively, and by using any of the following types of tests:
```
Reworded for clarity

sjhala-ccs commented on modules/virt-about-readiness-liveness-probes.adoc:
```diff
@@ -0,0 +1,19 @@
+
+Use readiness and liveness probes to detect and handle unhealthy virtual machine instances (VMIs). You can include one or more probes in the specification of the VMI to ensure that traffic does not reach a VMI that is not ready for it and that a new instance is created when a VMI becomes unresponsive.
+
+A _readiness probe_ determines whether a VMI is ready to accept service requests. If the probe fails, the VMI is removed from the list of available endpoints until the probe recovers.
```
Reworded for clarity

sjhala-ccs commented:
/cherrypick enterprise-4.6

sjhala-ccs commented:
/cherrypick enterprise-4.7

sjhala-ccs commented:
/cherrypick enterprise-4.8

sjhala-ccs commented:
/cherrypick enterprise-4.9
